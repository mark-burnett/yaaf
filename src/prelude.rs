pub use crate::{
    actor::{Actor, Tell},
    handler::{HandleContext, Handler},
    source::{Source, SourceContext},
    system::System,
};
pub use yaaf_macros::{Actor, Source};
