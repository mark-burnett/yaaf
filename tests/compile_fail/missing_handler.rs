use ::async_trait::async_trait;
use ::std::sync::Arc;
use ::tokio::sync::{
    oneshot::{channel, Sender},
    Mutex,
};
use ::yaaf::prelude::*;

#[derive(Clone)]
struct Communication(String);

#[derive(Actor)]
#[handle(Communication)]
struct Bob {
    done: Option<Sender<()>>,
    visited: Arc<Mutex<bool>>,
}

impl Bob {
    fn complete(&mut self) {
        let done = self.done.take();
        if let Some(done) = done {
            done.send(()).unwrap();
        }
    }
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn ::std::error::Error>> {
    let mut system = System::new().await;

    let (send, recv) = channel();
    let visited = Arc::new(Mutex::new(false));
    let bob = Bob {
        done: Some(send),
        visited: visited.clone(),
    };

    let bob_addr = system.add_actor(bob).await;

    bob_addr.tell(Communication("Hello".into()));

    recv.await.unwrap();
    system.shutdown().await;

    assert_eq!(true, *visited.lock().await);

    Ok(())
}
