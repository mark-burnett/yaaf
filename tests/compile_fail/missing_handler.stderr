warning: unused `std::result::Result` that must be used
   --> $DIR/router.rs:101:9
    |
101 | /         self.send.send(Envelope {
102 | |             distribution_type: DistributionType::Broadcast,
103 | |             message: message,
104 | |         });
    | |___________^
    |
    = note: `#[warn(unused_must_use)]` on by default
    = note: this `Result` may be an `Err` variant, which should be handled

warning: unused `std::result::Result` that must be used
   --> $DIR/router.rs:109:9
    |
109 | /         self.send.send(Envelope {
110 | |             distribution_type: DistributionType::Direct(recipient),
111 | |             message: message,
112 | |         });
    | |___________^
    |
    = note: this `Result` may be an `Err` variant, which should be handled

warning: unused `std::result::Result` that must be used
   --> $DIR/router.rs:118:9
    |
118 | /         self.sub_send
119 | |             .send(SubscriptionMessage::Subscribe((recipient, mailbox, s)));
    | |___________________________________________________________________________^
    |
    = note: this `Result` may be an `Err` variant, which should be handled

warning: unused `std::result::Result` that must be used
   --> $DIR/router.rs:177:41
    |
177 | ...                   send.send(envelope.message.clone());
    |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: this `Result` may be an `Err` variant, which should be handled

warning: unused `std::result::Result` that must be used
   --> $DIR/router.rs:180:41
    |
180 | ...                   send.send(envelope.message.clone());
    |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: this `Result` may be an `Err` variant, which should be handled

warning: unused `std::result::Result` that must be used
   --> $DIR/router.rs:184:37
    |
184 | ...                   self.subscribers.get(&recipient_id).unwrap().send(envelope.message);
    |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: this `Result` may be an `Err` variant, which should be handled

warning: 6 warnings emitted

warning: unused import: `::async_trait::async_trait`
 --> $DIR/missing_handler.rs:1:5
  |
1 | use ::async_trait::async_trait;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

error[E0277]: the trait bound `Bob: Handler<Communication>` is not satisfied
  --> $DIR/missing_handler.rs:12:10
   |
12 | #[derive(Actor)]
   |          ^^^^^ the trait `Handler<Communication>` is not implemented for `Bob`
   |
  ::: $WORKSPACE/src/actor.rs
   |
   | pub trait Actor: HandlesList<<Self as Actor>::Handles> {
   |                  ------------------------------------- required by this bound in `yaaf::Actor`
   |
   = note: required because of the requirements on the impl of `yaaf::handler::detail::HandlesList<(Communication,)>` for `Bob`
   = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)
